lib/dataconnection/DataConnection.ts

```diff
@@ -46,6 +59,20 @@ export abstract class DataConnection extends BaseConnection<
 	constructor(peerId: string, provider: Peer, options: any) {
 		super(peerId, provider, options);
 
+		this.then = (
+			onfulfilled?: (value: IDataConnection) => any,
+			onrejected?: (reason: PeerError<DataConnectionErrorType>) => any,
+		) => {
+			// Remove 'then' to prevent potential recursion issues
+			// `await` will wait for a Promise-like to resolve recursively
+			delete this.then;
```

Comment:
Doesn't this mean you cannot do e.g.

```javascript
const connectionPromise = peer.connect(peer);
connectionPromise.then(() => console.log(1));
connectionPromise.then(() => console.log(2));
```

and

```javascript
const connectionPromise = peer.connect(peer);
connectionPromise
  .then(() => console.log(1))
  .then(() => console.log(2))
  .catch(e => console.error(e))
```

? Also I think this will behave in an unexpected way, because the second `await` would finish before the connection is actually open:

```javascript
const connectionPromise = peer.connect(peer);
(async () => {
  await connectionPromise;
  console.log('connected');
})();
(async () => {
  await connectionPromise;
  console.log('connected');
  console.error('Sike! Not connected');
})();
```

While I think the first two examples are fine as long as it's clear to the users that it's not a perfect Promise implementation, the third one seems dangerous to me, it would be very hard to catch such a bug.

Here's a doc I found on how to make a good thenable: https://promisesaplus.com/#point-36

I think there must be an easy way to implement this by proxying `then` (and `catch`?) to a real `Promise` object.